fabric_generator/code_generator.py

```diff
@@ -0,0 +1,426 @@
+    def addDesignDescriptionEnd(self, indentLevel=0):
+        """
+        Add end of design description. 
+
+        Examples :
+            | Verilog: endmodule
+            | VHDL: end architecture Behavioral
+
+        Args:
+            indentLevel (int, optional): The indentation Level. Defaults to 0.
+        """
+        pass
+
+    @abc.abstractmethod
+    def addConstant(self, name: str, value, indentLevel=0):
+        """
+        Add a constant.
+
+        Examples :
+            | Verilog: parameter **name** = **value**;
+            | VHDL: constant **name** : STD_LOGIC := **value**;
+
+        Args:
+            name (str): name of the constant
+            value : The value of the constant.
+            indentLevel (int, optional): The indentation Level. Defaults to 0.
+        """
+        pass
+
+    @abc.abstractmethod
+    def addConnectionScalar(self, name: str, indentLevel=0):
+        """
+        Add a scalar connection.
+
+        Examples :
+            | Verilog: wire **name**;
+            | VHDL: signal **name** : STD_LOGIC;
+
+        Args:
+            name (str): name of the connection
+            indentLevel (int, optional): The indentation Level. Defaults to 0.
+        """
+        pass
+
+    @abc.abstractmethod
+    def addConnectionVector(self, name: str, startIndex, endIndex=0, indentLevel=0):
+        """
+        Add a vector connection.
+
+        Examples :
+            | Verilog: wire [**startIndex**:**end**] **name**;
+            | VHDL: signal **name** : STD_LOGIC_VECTOR( **startIndex** downto **endIndex** );
+
+        Args:
+            name (str): name of the connection
+            startIndex : Start index of the vector. Can be a string. 
+            endIndex (int, optional): End index of the vector. Can be a string. Defaults to 0.
+            indentLevel (int, optional): The indentation Level. Defaults to 0.
+        """
+        pass
+
+    @abc.abstractmethod
+    def addLogicStart(self, indentLevel=0):
+        """
+        Add start of logic. Only useful with VHDL.
+
+        Examples :
+            | Verilog: 
+            | VHDL: begin
+
+        Args:
+            indentLevel (int, optional): The indentation Level. Defaults to 0.
+        """
+        pass
+
+    @abc.abstractmethod
+    def addLogicEnd(self, indentLevel=0):
+        """
+        Add end of logic. Only useful with VHDL.
+
+        Examples :
+            | Verilog:
+            | VHDL: end
+
+        Args:
+            indentLevel (int, optional): The indentation Level. Defaults to 0.
+        """
+        pass
+
+    @abc.abstractmethod
+    def addInstantiation(self, compName: str, compInsName: str, compPorts: List[str], signals: List[str], paramPorts: List[str] = [], paramSignals: List[str] = [], indentLevel=0):
+        """
+        Add an instantiation. This will line up the ports and signals. So ports[0] will have signals[0] and so on. This is also the same case for paramPorts and paramSignals.
```

Comment:
sorry for dropping this so late in the day - but I think `dict`s, or perhaps a single array of `(port, sig)` tuples for `compPorts`/`signals` and `paramPorts`/`paramSignals` would be much neater and avoid the risk of these going out of sync when they're being built